File: explorer-agent/agent.py
#!/usr/bin/env python3

import json
import os
import sys
import uuid
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Optional
from urllib import error, parse, request


INTENTS = [
    "travel.booking.flight",
    "travel.booking.hotel",
    "food.delivery",
    "transport.ride_hailing",
    "software.purchase",
    "subscription.register",
    "freelance.hiring",
    "commerce.product_purchase",
    "finance.loan_application",
    "finance.insurance_quote",
]
DEFAULT_BASE_URL = "https://www.moltbillboard.com"
DEFAULT_INTENT = "software.purchase"
DEFAULT_LIMIT = 3


@dataclass
class Candidate:
    placement_id: str
    manifest: Dict[str, Any]
    offer: Dict[str, Any]
    score: int
    reasons: List[str]


def env(name: str, default: Optional[str] = None) -> Optional[str]:
    value = os.getenv(name)
    if value is None or value == "":
        return default
    return value


def parse_float(name: str, default: float) -> float:
    raw = env(name)
    if raw is None:
        return default
    try:
        return float(raw)
    except ValueError as exc:
        raise RuntimeError(f"Invalid {name}: {raw}") from exc


def api_request(
    base_url: str,
    method: str,
    path_or_url: str,
    payload: Optional[Dict[str, Any]] = None,
    headers: Optional[Dict[str, str]] = None,
) -> Any:
    url = path_or_url if path_or_url.startswith(("http://", "https://")) else parse.urljoin(base_url, path_or_url)
    data = None
    request_headers = {"Accept": "application/json"}

    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
        request_headers["Content-Type"] = "application/json"

    if headers:
        request_headers.update(headers)

    req = request.Request(url, data=data, headers=request_headers, method=method.upper())
    try:
        with request.urlopen(req, timeout=30) as response:
            raw = response.read().decode("utf-8")
            if not raw:
                return None
            return json.loads(raw)
    except error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"{method.upper()} {url} failed with {exc.code}: {body}") from exc
    except error.URLError as exc:
        raise RuntimeError(f"{method.upper()} {url} failed: {exc.reason}") from exc


def discover_placements(base_url: str, requested_intent: Optional[str], limit: int) -> Dict[str, Any]:
    intents: Iterable[str] = [requested_intent] if requested_intent else INTENTS

    for intent in intents:
        query = parse.urlencode({"intent": intent, "limit": limit})
        data = api_request(base_url, "GET", f"/api/v1/placements?{query}")
        placements = data.get("placements", [])
        if placements:
            return {"intent": intent, "placements": placements}

    raise RuntimeError("No live placements found for the requested discovery intents.")


def score_manifest_offer(manifest: Dict[str, Any], requested_intent: str) -> Candidate:
    placement = manifest["placement"]
    trust = placement.get("trust") or {}
    offers = placement.get("offers") or []
    placement_id = placement["id"]
    if not offers:
        raise RuntimeError(f"Placement {placement_id} did not expose any offers.")
    scored_offers: List[tuple[int, Dict[str, Any], List[str]]] = []
    for offer in offers:
        score = 0
        reasons: List[str] = []
        agent_hints = offer.get("agentHints") if isinstance(offer.get("agentHints"), dict) else {}

        if offer.get("primaryIntent") == requested_intent:
            score += 30
            reasons.append("offer matches requested intent")

        if offer.get("isPrimary"):
            score += 8
            reasons.append("offer is marked primary")

        if agent_hints.get("requiresAuth") is False:
            score += 4
            reasons.append("offer does not require auth")

        if agent_hints.get("expectedLatency") == "sync":
            score += 3
            reasons.append("offer is marked sync")

        if agent_hints.get("priceAvailable") is True:
            score += 2
            reasons.append("offer advertises price availability")

        scored_offers.append((score, offer, reasons))

    offer_score, best_offer, offer_reasons = max(
        scored_offers,
        key=lambda scored_offer: (scored_offer[0], scored_offer[1].get("offerId", "")),
    )

    score = offer_score
    reasons = list(offer_reasons)

    if trust.get("domainVerified"):
        score += 25
        reasons.append("placement passes homepage-to-destination domain verification")

    if trust.get("publisherVerified"):
        score += 15
        reasons.append("manifest is platform-signed")

    if trust.get("ownerTrustTier") == "trusted_internal":
        score += 15
        reasons.append("owner trust tier is trusted_internal")
    elif trust.get("ownerTrustTier") == "community_verified":
        score += 12
        reasons.append("owner trust tier is community_verified")
    elif trust.get("ownerTrustTier") == "email_verified":
        score += 8
        reasons.append("owner trust tier is email_verified")

    if trust.get("ownerVerificationStatus") == "homepage_verified":
        score += 10
        reasons.append("homepage proof-of-control completed")

    if trust.get("primaryDestinationStatus") == "verified_owner_domain":
        score += 10
        reasons.append("destination stays on verified owner domain")

    return Candidate(
        placement_id=placement_id,
        manifest=manifest,
        offer=best_offer,
        score=score,
        reasons=reasons,
    )


def report_action(
    base_url: str,
    action_id: str,
    placement_id: str,
    offer_id: str,
    event_type: str,
    intent: str,
) -> Dict[str, Any]:
    return api_request(
        base_url,
        "POST",
        "/api/v1/actions/report",
        payload={
            "actionId": action_id,
            "placementId": placement_id,
            "offerId": offer_id,
            "eventType": event_type,
            "metadata": {
                "source": "examples/explorer-agent",
                "intent": intent,
            },
        },
        headers={"Idempotency-Key": f"explorer-agent-{event_type}-{uuid.uuid4()}"},
    )


def report_conversion(
    base_url: str,
    action_id: str,
    placement_id: str,
    offer_id: str,
    conversion_type: str,
    value: float,
    currency: str,
    intent: str,
) -> Dict[str, Any]:
    return api_request(
        base_url,
        "POST",
        "/api/v1/conversions/report",
        payload={
            "actionId": action_id,
            "placementId": placement_id,
            "offerId": offer_id,
            "conversionType": conversion_type,
            "value": value,
            "currency": currency,
            "metadata": {
                "source": "examples/explorer-agent",
                "intent": intent,
            },
        },
    )


def env_bool(name: str, default: bool = False) -> bool:
    raw = env(name)
    if raw is None:
        return default
    return raw.strip().lower() in {"1", "true", "yes", "on"}


def main() -> int:
    base_url = env("MB_BASE", DEFAULT_BASE_URL)
    requested_intent = env("MB_INTENT", DEFAULT_INTENT)
    limit = max(1, int(env("MB_LIMIT", str(DEFAULT_LIMIT)) or DEFAULT_LIMIT))
    conversion_type = env("MB_CONVERSION_TYPE", "lead")
    conversion_value = parse_float("MB_CONVERSION_VALUE", 25.0)
    currency = env("MB_CURRENCY", "USD") or "USD"
    dry_run = env_bool("MB_DRY_RUN", default=False)

    print("MoltBillboard explorer agent")
    print(f"Base URL: {base_url}")
    print(f"Requested intent: {requested_intent}")
    print(f"Candidate limit: {limit}")
    print(f"Dry run: {dry_run}")

    discovered = discover_placements(base_url, requested_intent, limit)
    placements = discovered["placements"]
    print(f"Discovered {len(placements)} placement candidate(s)")

    candidates: List[Candidate] = []
    for placement in placements[:limit]:
        placement_id = placement["id"]
        manifest = api_request(base_url, "GET", f"/api/v1/placements/{placement_id}/manifest")
        candidate = score_manifest_offer(manifest, requested_intent)
        candidates.append(candidate)
        print(f"- {placement_id}: score={candidate.score}")

    chosen = max(candidates, key=lambda candidate: (candidate.score, candidate.placement_id))
    chosen_offer = chosen.offer
    chosen_attribution = chosen_offer.get("attribution") or {}
    action_id = chosen_attribution.get("actionId")
    offer_id = chosen_offer["offerId"]

    if not action_id:
        raise RuntimeError(f"Selected offer {offer_id} did not include a manifest-issued actionId.")

    print("\nSelected candidate")
    print(f"Placement: {chosen.placement_id}")
    print(f"Offer: {offer_id}")
    print(f"Action ID: {action_id}")
    print("Selection reasons:")
    for reason in chosen.reasons:
        print(f"  - {reason}")

    if dry_run:
        print("\nDry run complete (MB_DRY_RUN=1). Unset MB_DRY_RUN to report events.")
        return 0

    selected_result = report_action(
        base_url,
        action_id=action_id,
        placement_id=chosen.placement_id,
        offer_id=offer_id,
        event_type="offer_selected",
        intent=requested_intent,
    )
    print(f"\nReported offer_selected: {selected_result['success']}")

    executed_result = report_action(
        base_url,
        action_id=action_id,
        placement_id=chosen.placement_id,
        offer_id=offer_id,
        event_type="action_executed",
        intent=requested_intent,
    )
    print(f"Reported action_executed: {executed_result['success']}")

    conversion_result = report_conversion(
        base_url,
        action_id=action_id,
        placement_id=chosen.placement_id,
        offer_id=offer_id,
        conversion_type=conversion_type,
        value=conversion_value,
        currency=currency,
        intent=requested_intent,
    )
    print(f"Reported conversion: {conversion_result['success']}")

    stats_result = api_request(base_url, "GET", f"/api/v1/placements/{chosen.placement_id}/stats")
    stats = stats_result["stats"]
    print("\nStats snapshot")
    print(f"  offer_discovered: {stats['byType'].get('offer_discovered', 0)}")
    print(f"  offer_selected: {stats['byType'].get('offer_selected', 0)}")
    print(f"  action_executed: {stats['byType'].get('action_executed', 0)}")
    print(f"  conversion_reported: {stats['byType'].get('conversion_reported', 0)}")
    print(f"  conversion_count: {stats.get('conversionCount', 0)}")

    print("\nExplorer agent completed successfully.")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:  # noqa: BLE001
        print(f"Error: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: explorer-agent/agent.ts
/**
 * Explorer agent (TypeScript, no SDK) — intent-based placement browse →
 * manifest → attribution, plus an x402 payment step for the paid discovery
 * feed. Requires Node 18+. Run: npx tsx agent.ts
 *
 * The payment step needs its own dependencies (npm install in this folder)
 * and an AGENT_PRIVATE_KEY — see README.md. Without a key, it's skipped and
 * the rest of the loop still runs free.
 */

const INTENTS = [
  'travel.booking.flight',
  'travel.booking.hotel',
  'food.delivery',
  'transport.ride_hailing',
  'software.purchase',
  'subscription.register',
  'freelance.hiring',
  'commerce.product_purchase',
  'finance.loan_application',
  'finance.insurance_quote',
]

const DEFAULT_BASE_URL = 'https://www.moltbillboard.com'
const DEFAULT_INTENT = 'software.purchase'
const DEFAULT_LIMIT = 3

type Offer = {
  offerId: string
  primaryIntent?: string
  isPrimary?: boolean
  agentHints?: { requiresAuth?: boolean; expectedLatency?: string; priceAvailable?: boolean }
  attribution?: { actionId?: string }
}

type Manifest = {
  placement: {
    id: string
    trust?: {
      domainVerified?: boolean
      publisherVerified?: boolean
      ownerTrustTier?: string
      ownerVerificationStatus?: string
      primaryDestinationStatus?: string
    }
    offers?: Offer[]
  }
}

type Candidate = {
  placementId: string
  offer: Offer
  score: number
  reasons: string[]
}

function env(name: string, fallback?: string) {
  const v = process.env[name]
  return v == null || v === '' ? fallback : v
}

function parseFloatValue(name: string, fallback: number) {
  const raw = env(name)
  if (!raw) return fallback
  const parsed = Number.parseFloat(raw)
  if (!Number.isFinite(parsed)) throw new Error(`Invalid ${name}: ${raw}`)
  return parsed
}

async function apiRequest<T>(base: string, method: string, path: string, payload?: unknown, headers?: Record<string, string>): Promise<T> {
  const url = path.startsWith('http') ? path : new URL(path, base).toString()
  const res = await fetch(url, {
    method,
    headers: { Accept: 'application/json', ...(payload ? { 'Content-Type': 'application/json' } : {}), ...headers },
    body: payload ? JSON.stringify(payload) : undefined,
  })
  if (!res.ok) throw new Error(`${method} ${url} failed with ${res.status}: ${await res.text()}`)
  return (await res.json()) as T
}

async function discoverPlacements(base: string, requestedIntent: string | undefined, limit: number) {
  const intents = requestedIntent ? [requestedIntent] : INTENTS
  for (const intent of intents) {
    const query = new URLSearchParams({ intent, limit: String(limit) })
    const data = await apiRequest<{ placements: Array<{ id: string }> }>(base, 'GET', `/api/v1/placements?${query}`)
    if (data.placements.length > 0) return { intent, placements: data.placements }
  }
  throw new Error('No live placements found for the requested discovery intents.')
}

function scoreManifestOffer(manifest: Manifest, requestedIntent: string): Candidate {
  const placement = manifest.placement
  const trust = placement.trust || {}
  const offers = placement.offers || []
  if (offers.length === 0) throw new Error(`Placement ${placement.id} did not expose any offers.`)

  const scoredOffers = offers.map((offer) => {
    let score = 0
    const reasons: string[] = []
    const hints = offer.agentHints || {}

    if (offer.primaryIntent === requestedIntent) {
      score += 30
      reasons.push('offer matches requested intent')
    }
    if (offer.isPrimary) {
      score += 8
      reasons.push('offer is marked primary')
    }
    if (hints.requiresAuth === false) {
      score += 4
      reasons.push('offer does not require auth')
    }
    if (hints.expectedLatency === 'sync') {
      score += 3
      reasons.push('offer is marked sync')
    }
    if (hints.priceAvailable === true) {
      score += 2
      reasons.push('offer advertises price availability')
    }
    return { offer, score, reasons }
  })

  const bestOffer = scoredOffers.sort((a, b) => (b.score !== a.score ? b.score - a.score : a.offer.offerId.localeCompare(b.offer.offerId)))[0]

  let score = bestOffer.score
  const reasons = [...bestOffer.reasons]

  if (trust.domainVerified) {
    score += 25
    reasons.push('placement passes homepage-to-destination domain verification')
  }
  if (trust.publisherVerified) {
    score += 15
    reasons.push('manifest is platform-signed')
  }
  if (trust.ownerTrustTier === 'trusted_internal') {
    score += 15
    reasons.push('owner trust tier is trusted_internal')
  } else if (trust.ownerTrustTier === 'community_verified') {
    score += 12
    reasons.push('owner trust tier is community_verified')
  } else if (trust.ownerTrustTier === 'email_verified') {
    score += 8
    reasons.push('owner trust tier is email_verified')
  }
  if (trust.ownerVerificationStatus === 'homepage_verified') {
    score += 10
    reasons.push('homepage proof-of-control completed')
  }
  if (trust.primaryDestinationStatus === 'verified_owner_domain') {
    score += 10
    reasons.push('destination stays on verified owner domain')
  }

  return { placementId: placement.id, offer: bestOffer.offer, score, reasons }
}

async function reportAction(base: string, actionId: string, placementId: string, offerId: string, eventType: string, intent: string) {
  return apiRequest<{ success: boolean }>(
    base,
    'POST',
    '/api/v1/actions/report',
    { actionId, placementId, offerId, eventType, metadata: { source: 'explorer-agent/ts', intent } },
    { 'Idempotency-Key': `explorer-agent-${eventType}-${crypto.randomUUID()}` }
  )
}

async function reportConversion(base: string, actionId: string, placementId: string, offerId: string, conversionType: string, value: number, currency: string, intent: string) {
  return apiRequest<{ success: boolean }>(base, 'POST', '/api/v1/conversions/report', {
    actionId,
    placementId,
    offerId,
    conversionType,
    value,
    currency,
    metadata: { source: 'explorer-agent/ts', intent },
  })
}

// Pays for the x402-protected discovery feed with no account, no API key,
// and no sales call — just a funded wallet and a signature. Requires
// `npm install` in this folder first (see README.md) since @x402/fetch and
// @x402/evm aren't part of the zero-dependency scripts elsewhere in this repo.
async function payForDiscoveryFeed(base: string, maxDollars: number) {
  const raw = process.env.AGENT_PRIVATE_KEY || ''
  const key = raw.startsWith('0x') ? raw : raw ? `0x${raw}` : ''
  if (!/^0x[a-fA-F0-9]{64}$/.test(key)) {
    console.log('\nSkipping paid discovery feed (set AGENT_PRIVATE_KEY to a 0x… Base USDC wallet key to try it).')
    return
  }

  const { wrapFetchWithPaymentFromConfig } = await import('@x402/fetch')
  const { ExactEvmScheme } = await import('@x402/evm')
  const { privateKeyToAccount } = await import('viem/accounts')

  const account = privateKeyToAccount(key as `0x${string}`)
  const fetchWithPayment = wrapFetchWithPaymentFromConfig(fetch, {
    schemes: [{ network: 'eip155:8453', client: new ExactEvmScheme(account) }],
    // Explicit cap — @x402/core defaults to $1/payment if this is left unset.
    spendControls: { maxAmountPerPayment: `${maxDollars}` },
  })

  console.log('\nPaying for x402-protected discovery feed')
  console.log(`Wallet: ${account.address}`)
  const res = await fetchWithPayment(`${base}/api/x402/placements`)
  if (!res.ok) throw new Error(`Paid discovery feed request failed: ${res.status} ${await res.text()}`)
  const data = (await res.json()) as { placements?: unknown[] }
  console.log(`Paid ${data.placements?.length ?? 0} placement(s) from the discovery feed`)
}

async function main() {
  const base = env('MB_BASE', DEFAULT_BASE_URL)!
  const requestedIntent = env('MB_INTENT', DEFAULT_INTENT)!
  const limit = Math.max(1, Number.parseInt(env('MB_LIMIT', String(DEFAULT_LIMIT))!, 10) || DEFAULT_LIMIT)
  const conversionType = env('MB_CONVERSION_TYPE', 'lead')!
  const conversionValue = parseFloatValue('MB_CONVERSION_VALUE', 25)
  const currency = env('MB_CURRENCY', 'USD')!
  const dryRun = env('MB_DRY_RUN', '0') !== '0'

  console.log('MoltBillboard explorer agent (TS)')
  console.log(`Base URL: ${base}`)
  console.log(`Requested intent: ${requestedIntent}`)
  console.log(`Candidate limit: ${limit}`)
  console.log(`Dry run: ${dryRun}`)

  const discovered = await discoverPlacements(base, requestedIntent, limit)
  console.log(`Discovered ${discovered.placements.length} placement candidate(s)`)

  const candidates: Candidate[] = []
  for (const placement of discovered.placements.slice(0, limit)) {
    const manifest = await apiRequest<Manifest>(base, 'GET', `/api/v1/placements/${placement.id}/manifest`)
    const candidate = scoreManifestOffer(manifest, requestedIntent)
    candidates.push(candidate)
    console.log(`- ${placement.id}: score=${candidate.score}`)
  }

  const chosen = candidates.sort((a, b) => (b.score !== a.score ? b.score - a.score : a.placementId.localeCompare(b.placementId)))[0]
  const actionId = chosen.offer.attribution?.actionId
  if (!actionId) throw new Error(`Selected offer ${chosen.offer.offerId} did not include a manifest-issued actionId.`)

  console.log('\nSelected candidate')
  console.log(`Placement: ${chosen.placementId}`)
  console.log(`Offer: ${chosen.offer.offerId}`)
  console.log(`Action ID: ${actionId}`)
  console.log('Selection reasons:')
  for (const reason of chosen.reasons) console.log(`  - ${reason}`)

  if (dryRun) {
    console.log('\nDry run complete (MB_DRY_RUN=1). Unset MB_DRY_RUN to report events.')
    return
  }

  const selected = await reportAction(base, actionId, chosen.placementId, chosen.offer.offerId, 'offer_selected', requestedIntent)
  console.log(`\nReported offer_selected: ${selected.success}`)

  const executed = await reportAction(base, actionId, chosen.placementId, chosen.offer.offerId, 'action_executed', requestedIntent)
  console.log(`Reported action_executed: ${executed.success}`)

  const conversion = await reportConversion(base, actionId, chosen.placementId, chosen.offer.offerId, conversionType, conversionValue, currency, requestedIntent)
  console.log(`Reported conversion: ${conversion.success}`)

  const statsResult = await apiRequest<{ stats: { byType: Record<string, number>; conversionCount: number } }>(base, 'GET', `/api/v1/placements/${chosen.placementId}/stats`)
  const stats = statsResult.stats
  console.log('\nStats snapshot')
  console.log(`  offer_discovered: ${stats.byType.offer_discovered || 0}`)
  console.log(`  offer_selected: ${stats.byType.offer_selected || 0}`)
  console.log(`  action_executed: ${stats.byType.action_executed || 0}`)
  console.log(`  conversion_reported: ${stats.byType.conversion_reported || 0}`)
  console.log(`  conversion_count: ${stats.conversionCount || 0}`)

  await payForDiscoveryFeed(base, parseFloatValue('MB_X402_MAX_DOLLARS', 0.01))

  console.log('\nExplorer agent completed successfully.')
}

main().catch((error: unknown) => {
  const message = error instanceof Error ? error.message : 'Unknown error'
  console.error(`Error: ${message}`)
  process.exit(1)
})
